client/components/DebouncedTextField.jsx

```diff
@@ -0,0 +1,44 @@
+import React, { useCallback, useEffect, useState } from 'react';
+import PropTypes from 'prop-types';
+
+import { TextField } from '@material-ui/core';
+import { useDebouncedCallback } from 'use-debounce';
+
+const INPUT_DELAY = 200;
+
+export const DebouncedTextField = (props) => {
+  const [innerValue, setInnerValue] = useState('');
+
+  useEffect(() => {
+    if (props.value) {
+      setInnerValue(props.value);
+    } else {
+      setInnerValue('');
+    }
+  }, [props.value]);
+
+  const { callback: debouncedHandleOnChange } = useDebouncedCallback((event) => {
+    if (props.onChange) {
+      props.onChange(event);
+    }
+  }, INPUT_DELAY);
+
+  const handleOnChange = useCallback((event) => {
+    event.persist();
+
+    const newValue = event.currentTarget.value;
+    setInnerValue(newValue);
+    debouncedHandleOnChange(event);
+  }, []);
+
+  return <TextField {...props} value={innerValue} onChange={handleOnChange} />;
+};
+
+DebouncedTextField.propTypes = {
+  value: PropTypes.string,
+  onChange: PropTypes.func.isRequired,
+};
+
+DebouncedTextField.defaultProps = {
+  value: '',
+};
```

Comment:
I have no idea what this does but it looks great.

Reply:
This is to [`debounce` ](https://levelup.gitconnected.com/debounce-in-javascript-improve-your-applications-performance-5b01855e086) the text inputs - `<Formik />` has some performance issues around large forms of >50 fields, which is pretty easy for us to hit, so a debounce means that we only fire the `onChange` event every 200ms instead of on every change so that we get less rerenders

Reply:
interesting. thank you for the explanation.